File: week-2/readability.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <cs50.h>

int compute_grade(string text);
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
bool is_sentence_end(char letter);

int main(void)
{
    string text = get_string("Text: ");

    int grade = compute_grade(text);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }
}

int compute_grade(string text)
{
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // L is the average number of letters per 100 words
    float L = (float)(letters * 100) / words;
    // S is the average number of sentences per 100 words
    float S = (float)(sentences * 100) / words;
    //  Coleman-Liau index
    int index = round((0.0588 * L) - (0.296 * S) - 15.8);

    return index;
}

int count_letters(string text)
{
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        count += isalpha(text[i]) ? 1 : 0;
    }

    return count;
}

int count_words(string text)
{
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        count += isspace(text[i]) ? 1 : 0;

        if (i == strlen(text) - 1)
            count++;
    }

    return count;
}

int count_sentences(string text)
{
    int count = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        count += is_sentence_end(text[i]) ? 1 : 0;
    }

    return count;
}

bool is_sentence_end(char letter)
{
    return letter == '.' || letter == '!' || letter == '?';
}

File: week-6/readability.py
from cs50 import get_string


def main():
    text = get_string('Text: ')
    # computes the grade based on Coleman-Liau formula
    grade = compute_grade(text)
    # generates message based on the grade
    message = generate_message(grade)

    print(message)


def compute_grade(text):
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # L is the average number of letters per 100 words
    l = (letters * 100) / words
    # S is the average number of sentences per 100 words
    s = (sentences * 100) / words
    # Coleman-Liau index
    index = round((0.0588 * l) - (0.296 * s) - 15.8)

    return index


def count_letters(text):
    return len([c for c in text if c.isalpha()])


def count_words(text):
    return len([s for s in text if s.isspace()]) + 1


def count_sentences(text):
    punctiations = ['!', '.', '?']
    return len([p for p in text if p in punctiations])


def generate_message(grade):
    message = "Grade " + str(grade)

    # Changes message if grade is greater than 16
    if grade >= 16:
        message = "Grade 16+"
    # Changes message if grade is lesser than 16
    if grade < 1:
        message = "Before Grade 1"

    return message


if __name__ == '__main__':
    main()
